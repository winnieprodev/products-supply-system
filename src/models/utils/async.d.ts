export interface Error {
  message: string;
}

export interface Async<T> {
  isLoading: boolean;
  error: Error | undefined;
  data: T | undefined;
}
