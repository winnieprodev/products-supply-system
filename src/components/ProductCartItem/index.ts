export { default as ProductCartItem } from "./ProductCartItem";
