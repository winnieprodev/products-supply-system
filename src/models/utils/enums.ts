export enum WrapperMode {
  full = "full",
  singleLine = "singleLine",
  tiny = "tiny",
}
