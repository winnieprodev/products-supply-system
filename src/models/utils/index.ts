export * from "./async.d";
export * from "./enums";
