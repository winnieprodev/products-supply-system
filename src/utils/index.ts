export * from "./routeUtils";
export * from "./supplierUtils";
export * from "./testUtils";
