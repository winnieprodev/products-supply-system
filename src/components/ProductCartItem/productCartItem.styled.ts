import { GenericProductIcon } from "components/Icons";
import styled, { css, keyframes } from "styled-components";
import { secondaryColorLighter } from "styles/colors";

const removeItem = keyframes`
  from {
    opacity: 1;
    transform: translateX(0);
  }
  to {
    opacity: 0;
    transform: translateX(-300px);
  }
`;

interface WrapperProps {
  $isRemoved: boolean;
}
export const Wrapper = styled.div<WrapperProps>`
  display: flex;
  padding-bottom: 12px;
  ${({ $isRemoved }) =>
    $isRemoved &&
    css`
      animation: ${removeItem} 0.7s ease;
      animation-fill-mode: forwards;
    `};
`;

export const ImageBox = styled.div`
  width: 46px;
  height: 46px;
  background: ${secondaryColorLighter};
  border-radius: 10px;
`;

export const Image = styled.img`
  width: 100%;
  max-height: 100%;
  mix-blend-mode: darken;
`;

export const NoImage = styled(GenericProductIcon)`
  font-size: 20px;
  text-align: center;
  display: inherit;
  margin: 0 auto;
  padding: 12px 0;
  opacity: 0.2;
`;

export const Description = styled.div`
  width: 100%;
  padding: 0 8px;
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

export const Title = styled.div`
  font-weight: 600;
  font-size: 15px;
  padding-bottom: 4px;
`;

export const Quantities = styled.div`
  font-size: 12px;
`;

export const Icons = styled.div`
  display: flex;
  align-items: center;
  gap: 4px;
`;
