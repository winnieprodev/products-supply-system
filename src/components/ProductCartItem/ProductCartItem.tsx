import { AsyncWrapper, EditIcon, IconButton, TrashIcon } from "components";
import { ProductCartInfo, WrapperMode } from "models";
import { useGetProduct } from "queries";
import { useCallback, useState } from "react";
import { useHistory } from "react-router";
import ROUTES, { SupplierType } from "routes";

import * as Styled from "./productCartItem.styled";
interface OwnProps {
  item: ProductCartInfo;
  onRemove: () => void;
  outletId: string;
  supplierId: string;
  supplierType: SupplierType;
}

const ProductCartItem: React.FC<OwnProps> = ({
  item: { productId, quantities },
  onRemove,
  outletId,
  supplierId,
  supplierType,
}) => {
  const product = useGetProduct(productId);
  const [isRemoved, setIsRemoved] = useState(false);

  const history = useHistory();

  const onRemoveCallback = useCallback(() => {
    setIsRemoved(true);
    onRemove();
  }, [onRemove]);

  const onEditCallback = useCallback(() => {
    history.replace({
      pathname: ROUTES.SUPPLIER_DETAILS.use({
        productId,
        supplierId,
        supplierType,
        outletId,
      }),
      search: "?edit=true",
    });
  }, [history]);

  return (
    <AsyncWrapper item={product} mode={WrapperMode.singleLine}>
      {product.data && (
        <Styled.Wrapper $isRemoved={isRemoved}>
          <Styled.ImageBox>
            {product.data.image ? (
              <Styled.Image
                src={product.data.image}
                alt="Product image"
                aria-hidden
              />
            ) : (
              <Styled.NoImage />
            )}
          </Styled.ImageBox>
          <Styled.Description>
            <Styled.Title>{product.data.title}</Styled.Title>
            {quantities.map(({ type, amount }) => (
              <Styled.Quantities key={type}>
                {amount}{" "}
                {
                  product.data?.quantities.find(
                    (quantity) => quantity.type === type,
                  )?.name
                }
              </Styled.Quantities>
            ))}
          </Styled.Description>
          <Styled.Icons>
            <IconButton primary onClick={onEditCallback}>
              <EditIcon />
            </IconButton>
            <IconButton onClick={onRemoveCallback}>
              <TrashIcon />
            </IconButton>
          </Styled.Icons>
        </Styled.Wrapper>
      )}
    </AsyncWrapper>
  );
};

export default ProductCartItem;
