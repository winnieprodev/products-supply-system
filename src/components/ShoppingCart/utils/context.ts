import { Async, Cart, ProductCartInfo } from "models";
import React from "react";
import { mockAsync, mockCart } from "utils";

interface ShoppingCartContextProps {
  cartId: Async<number>;
  cart: Async<Cart> | undefined;
  setCart: (cart: Async<Cart> | undefined) => void;
  setProduct: (product: ProductCartInfo) => void;
  removeProduct: (productId: string) => void;
}

export const EMPTY_CART = { products: [] };
const mockFunction = () => {};

export const ShoppingCartContext = React.createContext<ShoppingCartContextProps>(
  {
    cartId: mockAsync(0),
    cart: mockAsync(mockCart(0)),
    setCart: mockFunction,
    setProduct: mockFunction,
    removeProduct: mockFunction,
  },
);
