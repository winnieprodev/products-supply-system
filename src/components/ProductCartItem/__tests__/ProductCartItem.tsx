import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { mockAsync } from "utils";

import ProductCartItem from "../ProductCartItem";

jest.mock("@apollo/client", () => ({
  gql: (req: string) => req,
  useQuery: () => () => mockAsync(undefined, true),
}));

describe("ProductCartItem", () => {
  describe("Snapshot tests", () => {
    test("it matches the snapshot", () => {
      const { container } = render(
        <BrowserRouter>
          <ProductCartItem
            item={{ productId: "id", quantities: [] }}
            onRemove={jest.fn()}
          />
        </BrowserRouter>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
