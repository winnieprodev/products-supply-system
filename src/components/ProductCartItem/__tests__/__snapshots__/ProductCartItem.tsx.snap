// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCartItem Snapshot tests it matches the snapshot 1`] = `null`;
